8291978: jpackage: allow to override primary l10n files on Windows

akashche opened the pull request:
This change is a follow-up to [this comment](https://bugs.openjdk.org/browse/JDK-8290519?focusedCommentId=14512038&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14512038).

Override implementation is based on [this comment](https://bugs.openjdk.org/browse/JDK-8290519?focusedCommentId=14510886&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14510886).

As suggested in [this comment](https://github.com/openjdk/jdk/pull/9753#pullrequestreview-1062560789) only English translation is provided for new `resource.wxl-file` label.

`getTempdirectory` utility was moved from `BasicTest` to `WindowsHelper` to be able to provide `--temp` and inspect the contents of `config` dir in verifier. It is not clear if `WindowsHelper` is an appropriate place for this utility (`BasicTest` is not Windows-specific), I've assumed that `--temp` flag itself is only used on Windows.

Testing:

 - [x] enhanced `WinL10nTest` adding checks for `-loc` arguments to `light.exe` and additional `@Parameter` run that overrides one of the primary `.wxl` files
 - [x] ran jtreg:jdk/tools/jpackage/windows and a `BasicTest` in unpack mode

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291978](https://bugs.openjdk.org/browse/JDK-8291978): jpackage: allow to override primary l10n files on Windows


### Reviewers
 * [Alexey Semenyuk](https://openjdk.org/census#asemenyuk) (@alexeysemenyukoracle - **Reviewer**)
 * [Alexander Matveev](https://openjdk.org/census#almatvee) (@sashamatveev - **Reviewer**) ⚠️ Review applies to [d4ba8bf0](https://git.openjdk.org/jdk/pull/9780/files/d4ba8bf0df3c519ff8e31700a16dd244d2bfb3d2)
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9780/head:pull/9780` \
`$ git checkout pull/9780`

Update a local copy of the PR: \
`$ git checkout pull/9780` \
`$ git pull https://git.openjdk.org/jdk pull/9780/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9780`

View PR using the GUI difftool: \
`$ git pr show -t 9780`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9780.diff">https://git.openjdk.org/jdk/pull/9780.diff</a>

</details>

akashche commented:
Thanks for the review!

akashche commented:
/integrate

sashamatveev commented on src/jdk.jpackage/windows/classes/jdk/jpackage/internal/resources/WinResources.properties:
```diff
@@ -34,6 +34,7 @@ resource.executable-properties-template=Template for creating executable propert
 resource.setup-icon=setup dialog icon
 resource.post-app-image-script=script to run after application image is populated
 resource.post-msi-script=script to run after msi file for exe installer is created
+resource.wxl-file=WiX localization file
```
This needs to be added to all WinResources*.properties files.

alexeysemenyukoracle replied:
Good point. We have a discussion of how l10n files should be changed at https://github.com/openjdk/jdk/pull/9753. Can you confirm that if some property is missing in a resource bundle its value is picked from the English resource bundle?

akashche replied:
I've just pushed the commit with EN entries for all files before reading the latest comment. I'll verify the behaviour with `.wxl` and `.properties` files and will comment in this issue.

naotoj replied:
I believe for the `.properties` case, modifications for the localized files are not required, as ResourceBundles fall back to the root (English) bundle.

alexeysemenyukoracle commented:
/sponsor

akashche commented:
Sorry for delay, I've found multiple problems while checking l10n handling manually.

Overview:

1. I've verified, that if an entry is missed in `WinResources_*.properties` file, fallback value from `WinResources.properties` file is actually used. Removed EN entries from non-EN `.properties` files in the patch.

2. I've found a problem in primary `.wxl` file selection, the [logic in code](https://github.com/openjdk/jdk/blob/ecfa38ffa8620e41854a043497f19863da297947/src/jdk.jpackage/windows/classes/jdk/jpackage/internal/WinMsiBundler.java#L567) uses `resource.wxl-file-name` to determine the file, but it appeared that all `WinResources_*.properties` files had the same `resource.wxl-file-name=MsiInstallerStrings_en.wxl` entry. I assumed this was an oversight and changed these entries to corresponding localized ones in the patch.

3. I've verified, that if an entry is missed in `.wxl` file, there is no fallback and `light.exe` invocation will fail like this:

```
InstallDirNotEmptyDlg.wxs(44) : error LGHT0102 : The localization variable !(loc.message.install.dir.exist) is unknown.  Please ensure the variable is defined.
```

This is caused because when `-J-Duser.language=de -J-Duser.country=DE ` are specified to `jpackage.exe `, it will result with the following cultures in `light.exe` invocation: `-cultures:de-de`.

4. Continuing with the problem in "3.", I've tried to add `en-us` culture as a fallback one, and found that it is not possible, `light.exe` fails with the following:

```
light.exe : error LGHT0100 : The localization identifier 'message.install.dir.exist' has been duplicated in multiple locations.  Please resolve the conflict.
```
when `-cultures:de-de;en-us` is specified to it.

5. `Culture` names in all primary `.wxl` files except `en-us` are specified using only "language" part, like `de`. I've found that this usage works correctly, when only `WixUtilExtension` is used (default jpackage behaviour), but it is not accepted by `light.exe` when `WixUIExtension` is also used, it fails like this (not Wix-internal build-time paths):

```
    C:\agent\_work\66\s\src\ext\UIExtension\wixlib\BrowseDlg.wxs(10) : error LGHT0102 : The localization variable !(loc.WixUIOK) is unknown.  Please ensure the variable is defined.
    C:\agent\_work\66\s\src\ext\UIExtension\wixlib\BrowseDlg.wxs(14) : error LGHT0102 : The localization variable !(loc.WixUICancel) is unknown.  Please ensure the variable is defined.
    C:\agent\_work\66\s\src\ext\UIExtension\wixlib\BrowseDlg.wxs(18) : error LGHT0102 : The localization variable !(loc.BrowseDlgComboLabel) is unknown.  Please ensure the variable is defined.
```
It is not clear, why exactly this is happening, I've found it is possible to fix this by using `Culture` names in primary `.wxl` files in exactly the same form as in Wix internal `.wxl` files, like in [WixUI_de-de.wxl](https://github.com/jef-n/wix3/blob/6b68e7f3601dfc35c3ee2a96510113585985b960/src/ext/UIExtension/wixlib/WixUI_de-de.wxl). I've updated cultures names in all primary `.wxl` files in the patch.

6. To cover "5." I've updated `WinL10nTest` adding runs with and without `WixUIExtension` for all primary cultures.

7. I assume this problem is not directly related to jpackage. I've found that `jpackage.exe` `--verbose` output for Japanese and Chinese locale runs is garbled when it is redirected to file:

```
[03:54:17.120] ?????? MSI ????? ???
```

I've found that if encoding is specified to std streams in -jpackage laucher here](https://github.com/openjdk/jdk/blob/37d3146cca2c40dd53fcebd9cb78595f018b3489/src/jdk.jpackage/share/classes/jdk/jpackage/main/Main.java#L50):

If these streams are changed to:

```
        PrintWriter out = new PrintWriter(System.out, false, StandardCharsets.UTF_8);
        PrintWriter err = new PrintWriter(System.err, false, StandardCharsets.UTF_8);
```

then redirected output in the file is correct. I am not sure whether this change makes sense (I understand that this is a console output, and console encoding handling will depend on Windows system locale, I used only `en-US` system locale), so I didn't include it in the patch. Maybe this is something to keep in mind for future with overall Windows move to better `UTF-8` support.

alexeysemenyukoracle commented:
Thank you for the deep investigation and detailed report!

#2 issue was an oversight in `WinResources_*.properties` files. Thank you for catching it!

#7 I'm 100% sure not specifying character encoding in PrintWriter ctor was an oversight. The solution with specifying utf8 encoding looks good to me. Please file a separate CR for this issue.

akashche commented:
Thanks for the review! I assume this PR still needs a l10n review from @naotoj before integrating it?

alexeysemenyukoracle commented:
Agree. It would be good if @naotoj will look at it on more time.

naotoj reviewed (approved):
LGTM. Sorry for the delay.

akashche commented:
/integrate

alexeysemenyukoracle commented:
/sponsor

akashche commented:
About the point "7." above, I've experimented (on Windows Server 2016) with `cmd.exe` and codepages and found that existing behaviour of `jpackage.exe` is consistent with existing behaviour of `java.exe`.

For both `java.exe` and `jpackage.exe`, when "Active codepage" in `cmd.exe` is `437` (default `en-US`):

```
C:\Users\Administrator>chcp
Active code page: 437
```

Then if I set default JVM locale to `ja-JP` using `-Duser.language=ja` and `-Duser.country=JP`, then both output in console window and redirected file output will be garbled to question marks `??????`.

If I set the Windows system "Language for non-Unicode programs" to Japanese, then, after the reboot, active codepage in `cmd.exe` will be 932 (note also Japanese file separator):

```
C:¥projects¥openjdk¥scripts>chcp
Active code page: 932
```

In this case running both `java.exe` and `jpackage.exe` with `ja-JP` JVM locale will cause Japanese output to be displayed correctly in the `cmd.exe`. And output redirected to file will also be written correctly in `Shift-JIS` encoding.

So I assume that existing behaviour is intended one and not filing the Jira issue now. Please let me know if it is still makes sense to file such Jira issue for `UTF-8` stdout (for example, for the future).

alexeysemenyukoracle commented:
I don't think it is good that the encoding of jpackage output depends on the default jvm encoding. It is convenient to know that the output is always in a specific encoding regardless of the values of `user.language` and `user.country` system properties. But this might not be the case of help and error output produced by jpackage.

Let's keep jpackage aligned with other JDK tools. Could you please check the behavior of `java`? How the output of `java --help` is encoded?

naotoj commented:
That's the expected behavior. Historically, we only support where user's locale and the system' locale match, otherwise such a garbled text would generate. Since Windows 10, they offer UTF-8 as the system encoding (still they claim it is beta), then all the out put is in UTF-8.

alexeysemenyukoracle commented:
Thank you for the clarification, @naotoj ! I think the case is closed. @akashche, no need to bother with #7.
